consistence/nsqd_coordinator_cluster_write.go

```diff
@@ -453,6 +455,261 @@ exitsync:
+		coordLog.Warningf("topic(%v) check write failed :%v", topicFullName, clusterWriteErr)
+		coordErrStats.incWriteErr(clusterWriteErr)
+		return clusterWriteErr
+	}
+	if isWrite && !tcData.IsISRReadyForWrite(self.myNode.GetID()) {
+		coordLog.Infof("topic(%v) operation failed since no enough ISR:%v", topicFullName, tcData.topicInfo)
+		coordErrStats.incWriteErr(ErrWriteQuorumFailed)
+		return ErrWriteQuorumFailed
+	}
+
+	checkCost := coordLog.Level() >= levellogger.LOG_DEBUG
+	if self.enableBenchCost {
+		checkCost = true
+	}
+
+	needRefreshISR := false
+	needLeaveISR := false
+	success := uint32(0)
+	failedNodes := make(map[string]struct{})
+	retryCnt := uint32(0)
+	exitErr := uint32(0)
+	var syncStart time.Time
+	var wg sync.WaitGroup
+
+	localErr := doLocalWrite(tcData)
+	if localErr != nil {
+		clusterWriteErr = localErr
+		goto exitsync
+	}
+	needLeaveISR = true
+
+	retrysync:
+	if retryCnt > MAX_WRITE_RETRY {
+		coordLog.Warningf("retrying times is large: %v", retryCnt)
+		needRefreshISR = true
+		if coord.IsExiting() {
+			clusterWriteErr = ErrTopicExiting
+			goto exitsync
+		}
+		time.Sleep(time.Second)
+	}
+	if needRefreshISR {
+		tcData = coord.GetData()
+		if clusterWriteErr = tcData.checkWriteForLeader(self.myNode.GetID()); clusterWriteErr != nil {
+			coordLog.Warningf("topic(%v) check operation failed :%v", topicFullName, clusterWriteErr)
+			goto exitsync
+		}
+		if clusterWriteErr = doRefresh(tcData); clusterWriteErr != nil {
+			coordLog.Warningf("topic(%v) failed refresh data:%v", topicFullName, clusterWriteErr)
+			goto exitsync
+		}
+		if isWrite && !tcData.IsISRReadyForWrite(self.myNode.GetID()) {
+			coordLog.Infof("topic(%v) sync write failed since no enough ISR:%v", topicFullName, tcData.topicInfo)
+			coordErrStats.incWriteErr(ErrWriteQuorumFailed)
+			clusterWriteErr = ErrWriteQuorumFailed
+			goto exitsync
+		}
+		if retryCnt > 3 {
+			go self.requestNotifyNewTopicInfo(topicName, topicPartition)
+		}
+	}
+	success = 0
+	retryCnt++
+
+	// send message to slaves with current topic epoch
+	// replica should check if offset matching. If not matched the replica should leave the ISR list.
+	// also, the coordinator should retry on fail until all nodes in ISR success.
+	// If failed, should update ISR and retry.
+	// write epoch should keep the same (ignore epoch change during write)
+	exitErr = 0
+	for _, nodeID := range tcData.topicInfo.ISR {
+		if nodeID == self.myNode.GetID() {
+			success++
+			continue
+		}
+
+		c, rpcErr := self.acquireRpcClient(nodeID)
+		if rpcErr != nil {
+			coordLog.Infof("get rpc client %v failed: %v", nodeID, rpcErr)
+			needRefreshISR = true
+			failedNodes[nodeID] = struct{}{}
+			continue
+		}
+
+		// should retry if failed, and the slave should keep the last success write to avoid the duplicated
+		retSlave, rpcErr := doSlaveSync(c, nodeID, tcData)
+		if rpcErr == nil {
+			wg.Add(1)
+			id := nodeID
+			go func(){
```

Comment:
I think no goroutinues should be used here, just wait on the slave result channel is enough